firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp

```diff
@@ -0,0 +1,391 @@
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element is initalised with payload 0x01") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    input_payload.push_back(0x01);
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("the crc is computed from the payload") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0xa016d052;
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      output_payload.resize(TestCRCElementHeaderProps::header_size + input_payload.size());
+      output_payload.copy_from(
+        input_payload.buffer(), input_payload.size(), TestCRCElementHeaderProps::payload_offset);
+
+      auto computed_crc = crc_element.compute_body_crc(output_payload, crc32c);
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    auto data = PF::Util::make_array<uint8_t>(0x01, 0x02, 0x05);
+    for (auto& bytes : data) {
+      input_payload.push_back(bytes);
+    }
+
+    TestCRCElement crc_element{input_payload};
+    WHEN("given payload data is written to it") {
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      auto write_status = crc_element.write(output_payload, crc32c);
+      auto crc = crc_element.crc();
+      auto crc_compute = crc32c.compute(input_payload.buffer(), input_payload.size());
+
+      THEN("crc computed in write method is same as crcchecker compute") {
+        REQUIRE(write_status == PF::IndexStatus::ok);
+        REQUIRE(crc == crc_compute);
+      }
+    }
+  }
+
+  GIVEN("crc element is initalised with 0 bytes") {
+
+    TestCRCElementProps::PayloadBuffer input_payload;
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("given payload data is written to it") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0x6D551F4C;
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+
+      auto write_status = crc_element.write(input_buffer, crc32c);
+      auto crc_payload = crc_element.crc();
+      THEN("CRC is equal to 0") {
+        REQUIRE(crc_payload == 0);
+      }
+    }
+  }
+
+  GIVEN("A CRC element is initalised with valid payload parameter") {
+    constexpr size_t buffer_size = 254UL;
+    using TestCRCElementProps = PF::Protocols::CRCElementProps<buffer_size>;
+    using TestCRCElement = PF::Protocols::CRCElement<TestCRCElementProps::PayloadBuffer>;
+    PF::HAL::SoftCRC32 crc32c{PF::HAL::crc32c_params};
+
+    auto expected_payload = std::string("\x01\x05\x01\x02\x03\x04\x05", 7);
+    auto input_data = PF::Util::make_array<uint8_t>(0x01, 0x05, 0x01, 0x02, 0x03, 0x04, 0x05);
+    TestCRCElementProps::PayloadBuffer input_payload;
```

Comment:
I think the variable naming `input_payload` is misleading here: we're really just giving a buffer for the CRC to write the payload to when it parses the body (which is the `input_data` variable). So this `input_payload` isn't being used as an input when we're parsing the body.

Reply:
Right! actually I realised i can simply initialise an empty payload buffer here instead of giving it some 'input_data' , therefor changing the variable name to just 'payload'